backend/src/game/game.service.ts

```diff
@@ -23,30 +25,298 @@ export class GameService {
+    if (pendingPlayer) {
+      try {
+        const pendingPlayerObject: Player = JSON.parse(pendingPlayer);
+        if (pendingPlayerObject.email === playerEmail) {
+          await this.cacheManager.del('pendingPlayer');
+          await this.cacheManager.del(`game${playerEmail}`);
+          await this.cacheManager.del(pendingPlayerObject.gameID);
+        }
+      } catch (error) {
+        console.log(error);
+      }
+    }
   }
 
-  async joinOrCreateGame(player: Player): Promise<[boolean, string]> {
+  async updateUserDisconnectStatus(client: Socket) {
+    // find the user in database
+    const user = await this.getSocketUser(client);
+    if (!user) {
+      return;
+    }
+    if (user.status === 'WAITING' || user.status === 'PLAYING') {
+      try {
+        await this.prisma.user.update({
+          where: {
+            email: user.email,
+          },
+          data: {
+            status: 'AWAY',
+          },
+        });
+      } catch (err) {
+        console.log(err);
+      }
+    }
+  }
+
+  async clearData(client: Socket) {
+    console.log('cleaning!!');
+    const player = await this.getSocketPlayer(client);
+    if (!player) return;
+
+    const game: Game = await this.getGameByID(player.gameID);
+    if (game && game.status === GameStatus.Ended) {
+      await this.cacheManager.del(`game${game.gameID}`);
+    }
+    if (player.gameID === 'non') {
+      await this.cacheManager.del(`game${player.email}`);
+    } else {
+      await this.pauseGame(player.gameID);
+    }
+    await this.cacheManager.del(client.id);
+  }
+
+  async pauseGame(gameID: string) {
+    const game = await this.getGameByID(gameID);
+    if (!game) return;
+    game.status = GameStatus.Pause;
+    await this.cacheManager.set(gameID, JSON.stringify(game));
+  }
+
+  /****************************************************************************/
+  // Create, Update Player(before finding a game)
+  /****************************************************************************/
+  async getPlayerByEmail(email: string): Promise<Player> {
+    let player: Player;
+    const playerString: string = await this.cacheManager.get(`game${email}`);
+    if (playerString) { 
+      try {
+        player = JSON.parse(playerString);
+      } catch(error) {
+        console.log(error);
+      }
+    }
+    return player;
+  }
+
+  async getSocketPlayer(client: Socket) {
+    let player: Player;
+    const playerEmail: string = await this.cacheManager.get(client.id);
+    const playerString: string = await this.cacheManager.get(
+      `game${playerEmail}`,
+    );
+    if (playerString) player = JSON.parse(playerString);
+    return player;
+  }
+
+  async createPlayer(client: Socket): Promise<Player> {
+    const pausingPlayer = await this.updatePausingPlayer(client);
+    console.log("has pausing player: " , pausingPlayer);
+    if (pausingPlayer) return pausingPlayer;
+
+    const user = await this.getSocketUser(client);
```

Comment:
Can you add typing? Because we have prisma users, game users, chat users.

Reply:
Should add typing!

Here in the game.service I'm using "User" to refer to the "user in database", and "Player" to refer to the "user playing the game". But we can just make a unification like "User", "Player", "ChatUser" for three different user types??

Reply:
I meant typing as such: `const user: Player` in stead of `const user`
(so we don't have to look at the return of the function to see what type we are getting)